File: c programming/sorting.c
// #include <stdio.h>
// void bubble(int arr[], int n)
// {
//     for (int i = 0; i < n - 1; i++)
//     {
//         for (int j = 0; j < n - i - 1; j++)
//         {
//             if (arr[j + 1] < arr[j])
//             {
//                 int temp = arr[j];
//                 arr[j] = arr[j + 1];
//                 arr[j + 1] = temp;
//             }
//         }
//     }
//     for (int i = 0; i < n; i++)
//     {
//         printf("%d ", arr[i]);
//     }
// }
// int main()
// {
//     int arr[] = {2, 5, 3, 6, 2, 1};
//     int n = sizeof(arr) / sizeof(arr[0]);
//     bubble(arr, n);
// }

// #include <stdio.h>
// void selection(int arr[], int n)
// {
//     for (int i = 0; i < n - 1; i++)
//     {
//         int min = i;
//         for (int j = i + 1; j < n; j++)
//         {
//             if (arr[j] < arr[min])
//             {
//                 min = j;
//             }
//         }
//         int temp = arr[i];
//         arr[i] = arr[min];
//         arr[min] = temp;
//     }
//     for (int i = 0; i < n; i++)
//     {
//         printf("%d ",arr[i]);
//     }
    
// }
// int main()
// {
//     int arr[] = {2, 5, 3, 6, 2, 1};
//     int n = sizeof(arr) / sizeof(arr[0]);
//     selection(arr, n);
// }

// #include <stdio.h>
// void insertion(int arr[], int n)
// {
//     for (int i = 1; i < n; i++)
//     {
//         int curr = arr[i];
//         int j = i-1;
//         while (j>=0 && curr<arr[j])
//         {
//             arr[j+1] = arr[j];
//             j--;
//         }
//         arr[j+1] = curr;
//     }
//     for (int i = 0; i < n; i++)
//     {
//         printf("%d ",arr[i]);
//     }
    
    
// }
// int main()
// {
//     int arr[] = {2, 5, 3, 6, 2, 1};
//     int n = sizeof(arr) / sizeof(arr[0]);
//     insertion(arr,n);
// }
// #include<stdio.h>
// void mergesort(int arr[], int si, int ei){
//     if (si>=ei)
//     {
//         return;
//     }
//     int mid = si+(ei-si)/2;
//     mergesort(arr,si,mid);
//     mergesort(arr,mid+1,ei);

// }
// void merger(int arr[], int si, int mid, int ei, int n){
//     int temp[ei-si+1];
//     int i =si;
//     int j = mid +1;
//     int k = 0;
//     while (i <= mid && j <= ei){
//         if (arr[i]<arr[j])
//         {
//             temp[k]=arr[i];
//             i++;
//         }else{
//             temp[k]= arr[j];
//             j++;
//         }
//         k++;
//     }
//     while (i<=mid)
//     {
//         temp[k++]= arr[i++];
//     }
//     while (j<=ei)
//     {
//        temp[k++] = arr[j++];
//     }
//     for (k=0, i=si; i <n ; k++, i++)
//     {
//         arr[i] = temp[k];
//     }
// }


#include <stdio.h>

void bubbleSort(int arr[], int n) {
    int i, j, temp;
    for (i = 0; i < n-1; i++) {
        for (j = 0; j < n-i-1; j++) {
            if (arr[j] > arr[j+1]) {
                // Swap the elements
                temp = arr[j];
                arr[j] = arr[j+1];
                arr[j+1] = temp;
            }
        }
    }
}

int main() {
    int arr[] = {64, 25, 12, 22, 11};
    int n = sizeof(arr)/sizeof(arr[0]);
    
    bubbleSort(arr, n);
    
    printf("Sorted array: ");
    for (int i = 0; i < n; i++)
        printf("%d ", arr[i]);
    return 0;
}

File: python programming/List/sorting.py
# def bubble_sort(arr):
#     n = len(arr)

#     for i in range(n):
#         for j in range(0, n - i - 1):
#             if arr[j] > arr[j + 1]:
#                 arr[j], arr[j + 1] = arr[j + 1], arr[j]
#     return arr
# arr = [5, 3, 8, 4, 2]
# print(bubble_sort(arr))


def merge(arr, left, mid , right):
    l = arr[left:mid + 1]
    r = arr[mid+1:right+1]

    i = 0
    j = 0
    k = left

    while i < len(l) and j < len(r):
        if l[i] <= r[j]:
            arr[k] = l[i]
            i += 1
        else:
            arr[k] = arr[j]
            j += 1
        k += 1
    while i < len(l):
        arr[k] = l[i]
        i += 1
        k += 1
    while j < len(r):
        arr[k] = r[j]
        j += 1
        k += 1
    
def merge_sort(arr, left, right):
    if left < right:
        mid = (left + right) // 2

        merge_sort(arr, left, mid)
        merge_sort(arr, mid + 1, right)

        merge(arr, left, mid, right)

def main():
    arr = [38, 27, 43, 3, 9, 82, 10]

    merge_sort(arr, 0, len(arr) - 1)

    print("Sorted array:", arr)


if __name__ == "__main__":
    main()